src/utils/utils.js

```diff
@@ -100,28 +100,53 @@ function Utils() {
    */
   this.trackHeader = gaUtils.trackEvent('Global Header');
 
+  /**
+   * createFunctionWithTimeout(callback, optTimeout)
+   * The function serves as a pipe to return the function that is passed to it.
+   * It also searves as a timer to execute that function after a certain amount of time.
+   *
+   * @param {Function} callback - The function to be executed after the time of optTimeout
+   * @param {Number} optTimeout
+   * @return {Function}
+   */
+  this.createFunctionWithTimeout = (callback, optTimeout) => {
+  let called = false;
+
+  const fn = () => {
+    if (!called) {
+      called = true;
+      callback();
+    }
+  };
+
+  setTimeout(fn, optTimeout || 500);
+
+  return fn;
```

Comment:
Do you need to return the function here if you're going to call it in a setTimeout? You should also add a clearTimeout to all setTimeouts.

Reply:
It is for GA. The codes is from the what GA's instruction. Basically, the reason why we need to return a function here is for the case once a GA event is not successfully sent through, the app wound't wait forever but just go on the execute the original event, such as a click. So it doesn't need to clearTimeout in this scenario as it will be execute after all, not matter a GA event is sent or not.

Ref, See the "Handling Timeouts" Section!
https://developers.google.com/analytics/devguides/collection/analyticsjs/sending-hits